# Elevate Labs - Data Analyst Internship

## Task 1: Data Cleaning and Preprocessing

### Dataset: Customer Personality Analysis

**Tools Used:**
- Python
- Pandas
- NumPy
- Google Colab

### Objective
Clean and preprocess a raw customer dataset by identifying and handling missing values, duplicate records, inconsistent formats, incorrect data types, and other data-quality issues.


In [105]:
import pandas as pd
import numpy as np

print("Pandas version:", pd.__version__)
print("NumPy version:", np.__version__)

Pandas version: 2.2.3
NumPy version: 2.1.3


In [106]:
df = pd.read_csv("/content/marketing_campaign.csv", sep="\t")

print("Dataset loaded successfully!")

Dataset loaded successfully!


In [107]:
df.head()

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,04-09-2012,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,08-03-2014,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,21-08-2013,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,10-02-2014,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,19-01-2014,94,173,...,5,0,0,0,0,0,0,3,11,0


In [108]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])

Number of rows: 2240
Number of columns: 29


In [109]:
df.shape

(2240, 29)

In [110]:
print(df.columns.tolist())

['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response']


In [111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   int64  
 1   Year_Birth           2240 non-null   int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   int64  
 6   Teenhome             2240 non-null   int64  
 7   Dt_Customer          2240 non-null   object 
 8   Recency              2240 non-null   int64  
 9   MntWines             2240 non-null   int64  
 10  MntFruits            2240 non-null   int64  
 11  MntMeatProducts      2240 non-null   int64  
 12  MntFishProducts      2240 non-null   int64  
 13  MntSweetProducts     2240 non-null   int64  
 14  MntGoldProds         2240 non-null   int64  
 15  NumDealsPurchases    2240 non-null   i

In [112]:
df.describe()

,ID,Year_Birth,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
count,2240.000000,2240.000000,2216.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,...,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.0,2240.0,2240.000000
mean,5592.159821,1968.805804,52247.251354,0.444196,0.506250,49.109375,303.935714,26.302232,166.950000,37.525446,...,5.316518,0.072768,0.074554,0.072768,0.064286,0.013393,0.009375,3.0,11.0,0.149107
std,3246.662198,11.984069,25173.076661,0.538398,0.544538,28.962453,336.597393,39.773434,225.715373,54.628979,...,2.426645,0.259813,0.262728,0.259813,0.245316,0.114976,0.096391,0.0,0.0,0.356274
min,0.000000,1893.000000,1730.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
25%,2828.250000,1959.000000,35303.000000,0.000000,0.000000,24.000000,23.750000,1.000000,16.000000,3.000000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
50%,5458.500000,1970.000000,51381.500000,0.000000,0.000000,49.000000,173.500000,8.000000,67.000000,12.000000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
75%,8427.750000,1977.000000,68522.000000,1.000000,1.000000,74.000000,504.250000,33.000000,232.000000,50.000000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
max,11191.000000,1996.000000,666666.000000,2.000000,2.000000,99.000000,1493.000000,199.000000,1725.000000,259.000000,...,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,3.0,11.0,1.000000


In [113]:
df.describe(include="all")

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
count,2240.000000,2240.000000,2240,2240,2216.000000,2240.000000,2240.000000,2240,2240.000000,2240.000000,...,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.0,2240.0,2240.000000
unique,NaN,NaN,5,8,NaN,NaN,NaN,663,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,NaN,Graduation,Married,NaN,NaN,NaN,31-08-2012,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,NaN,1127,864,NaN,NaN,NaN,12,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,5592.159821,1968.805804,NaN,NaN,52247.251354,0.444196,0.506250,NaN,49.109375,303.935714,...,5.316518,0.072768,0.074554,0.072768,0.064286,0.013393,0.009375,3.0,11.0,0.149107
std,3246.662198,11.984069,NaN,NaN,25173.076661,0.538398,0.544538,NaN,28.962453,336.597393,...,2.426645,0.259813,0.262728,0.259813,0.245316,0.114976,0.096391,0.0,0.0,0.356274
min,0.000000,1893.000000,NaN,NaN,1730.000000,0.000000,0.000000,NaN,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
25%,2828.250000,1959.000000,NaN,NaN,35303.000000,0.000000,0.000000,NaN,24.000000,23.750000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
50%,5458.500000,1970.000000,NaN,NaN,51381.500000,0.000000,0.000000,NaN,49.000000,173.500000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
75%,8427.750000,1977.000000,NaN,NaN,68522.000000,1.000000,1.000000,NaN,74.000000,504.250000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000


In [114]:
missing_values = df.isnull().sum()

print(missing_values)

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64


In [115]:
missing_report = pd.DataFrame({
    "Missing Values": df.isnull().sum(),
    "Missing Percentage": (
        df.isnull().sum() / len(df) * 100
    ).round(2)
})

missing_report = missing_report[
    missing_report["Missing Values"] > 0
]

missing_report

,Missing Values,Missing Percentage
Income,24,1.07


In [116]:
duplicate_count = df.duplicated().sum()

print("Duplicate rows:", duplicate_count)

Duplicate rows: 0


In [117]:
df[df.duplicated()]

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response


In [118]:
print("Duplicate customer IDs:", df["ID"].duplicated().sum())

Duplicate customer IDs: 0


In [119]:
original_columns = df.columns.tolist()

print("Original columns:")
print(original_columns)

Original columns:
['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits', 'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts', 'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases', 'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth', 'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1', 'AcceptedCmp2', 'Complain', 'Z_CostContact', 'Z_Revenue', 'Response']


In [120]:
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

In [121]:
print("Cleaned columns:")
print(df.columns.tolist())

Cleaned columns:
['id', 'year_birth', 'education', 'marital_status', 'income', 'kidhome', 'teenhome', 'dt_customer', 'recency', 'mntwines', 'mntfruits', 'mntmeatproducts', 'mntfishproducts', 'mntsweetproducts', 'mntgoldprods', 'numdealspurchases', 'numwebpurchases', 'numcatalogpurchases', 'numstorepurchases', 'numwebvisitsmonth', 'acceptedcmp3', 'acceptedcmp4', 'acceptedcmp5', 'acceptedcmp1', 'acceptedcmp2', 'complain', 'z_costcontact', 'z_revenue', 'response']


In [122]:
print("Education values:")
print(df["education"].value_counts(dropna=False))

Education values:
education
Graduation    1127
PhD            486
Master         370
2n Cycle       203
Basic           54
Name: count, dtype: int64


In [123]:
print("Marital status values:")
print(df["marital_status"].value_counts(dropna=False))

Marital status values:
marital_status
Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
Absurd        2
YOLO          2
Name: count, dtype: int64


### Categorical Data Validation

The Education and Marital_Status columns were inspected for inconsistent values and formatting.

Unnecessary whitespace was removed from text fields. Existing category labels were preserved because the dataset alone does not provide sufficient evidence to determine whether unusual category labels should be merged or renamed.

In [124]:
text_columns = df.select_dtypes(include=["object"]).columns

for col in text_columns:
    df[col] = df[col].str.strip()

In [125]:
print("Missing Income before cleaning:",
      df["income"].isnull().sum())

Missing Income before cleaning: 24


In [126]:
income_median = df["income"].median()

print("Income median:", income_median)

Income median: 51381.5


In [127]:
df["income"] = df["income"].fillna(income_median)

In [128]:
print("Missing Income after cleaning:",
      df["income"].isnull().sum())

Missing Income after cleaning: 0


In [129]:
df["dt_customer"].head()

,dt_customer
0,04-09-2012
1,08-03-2014
2,21-08-2013
3,10-02-2014
4,19-01-2014


In [130]:
df["dt_customer"] = pd.to_datetime(
    df["dt_customer"],
    dayfirst=True,
    errors="coerce"
)

In [131]:
print(df["dt_customer"].dtype)

datetime64[ns]


In [132]:
print(
    "Invalid dates:",
    df["dt_customer"].isnull().sum()
)

Invalid dates: 0


In [133]:
numeric_columns = [
    "year_birth",
    "income",
    "kidhome",
    "teenhome",
    "recency",
    "mntwines",
    "mntfruits",
    "mntmeatproducts",
    "mntfishproducts",
    "mntsweetproducts",
    "mntgoldprods"
]

In [134]:
for col in numeric_columns:
    df[col] = pd.to_numeric(df[col], errors="coerce")

In [135]:
df[numeric_columns].dtypes

,0
year_birth,int64
income,float64
kidhome,int64
teenhome,int64
recency,int64
mntwines,int64
mntfruits,int64
mntmeatproducts,int64
mntfishproducts,int64
mntsweetproducts,int64


In [136]:
print("Negative income:",
      (df["income"] < 0).sum())

print("Negative recency:",
      (df["recency"] < 0).sum())

print("Negative wine spending:",
      (df["mntwines"] < 0).sum())

print("Negative fruit spending:",
      (df["mntfruits"] < 0).sum())

Negative income: 0
Negative recency: 0
Negative wine spending: 0
Negative fruit spending: 0


In [137]:
# Check for potential Income outliers using the IQR method

Q1 = df["income"].quantile(0.25)
Q3 = df["income"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

income_outliers = df[
    (df["income"] < lower_bound) |
    (df["income"] > upper_bound)
]

print("Lower bound:", lower_bound)
print("Upper bound:", upper_bound)
print("Potential Income outliers:", len(income_outliers))

income_outliers[["id", "income"]].head(10)

Lower bound: -13587.75
Upper bound: 117416.25
Potential Income outliers: 8


,id,income
164,8475,157243.0
617,1503,162397.0
655,5555,153924.0
687,1501,160803.0
1300,5336,157733.0
1653,4931,157146.0
2132,11181,156924.0
2233,9432,666666.0


### Outlier Check

Potential Income outliers were identified using the Interquartile Range (IQR) method. These observations were not automatically removed because an outlier can represent a genuine customer value rather than a data-entry error. Further business context would be required before removing or modifying such records.

In [138]:
print("Minimum birth year:", df["year_birth"].min())
print("Maximum birth year:", df["year_birth"].max())

Minimum birth year: 1893
Maximum birth year: 1996


In [139]:
# Check for potentially unusual birth years

current_reference_year = 2014

age_check = current_reference_year - df["year_birth"]

print("Minimum calculated age:", age_check.min())
print("Maximum calculated age:", age_check.max())

print("\nCustomers with calculated age above 100:")
df.loc[age_check > 100, ["id", "year_birth"]]

Minimum calculated age: 18
Maximum calculated age: 121

Customers with calculated age above 100:


,id,year_birth
192,7829,1900
239,11004,1893
339,1150,1899


### Birth Year Data Quality Check

The `Year_Birth` column was inspected for potentially unusual values.

Some very old birth years were identified in the dataset. These records were not automatically removed because an unusual value does not necessarily mean that it is incorrect. The records were retained because there was insufficient evidence within the dataset to confirm that they were data-entry errors.

In [140]:
final_missing = df.isnull().sum()

print(final_missing)

id                     0
year_birth             0
education              0
marital_status         0
income                 0
kidhome                0
teenhome               0
dt_customer            0
recency                0
mntwines               0
mntfruits              0
mntmeatproducts        0
mntfishproducts        0
mntsweetproducts       0
mntgoldprods           0
numdealspurchases      0
numwebpurchases        0
numcatalogpurchases    0
numstorepurchases      0
numwebvisitsmonth      0
acceptedcmp3           0
acceptedcmp4           0
acceptedcmp5           0
acceptedcmp1           0
acceptedcmp2           0
complain               0
z_costcontact          0
z_revenue              0
response               0
dtype: int64


In [141]:
print(
    "Total missing values:",
    df.isnull().sum().sum()
)

Total missing values: 0


In [142]:
print(
    "Duplicate rows after cleaning:",
    df.duplicated().sum()
)

Duplicate rows after cleaning: 0


In [143]:
duplicate_count = df.duplicated().sum()

if duplicate_count > 0:
    df = df.drop_duplicates()
    print(f"Removed {duplicate_count} duplicate rows.")
else:
    print("No duplicate rows found. No rows removed.")

No duplicate rows found. No rows removed.


In [144]:
print(
    "Duplicate rows after removal:",
    df.duplicated().sum()
)

Duplicate rows after removal: 0


In [145]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   id                   2240 non-null   int64         
 1   year_birth           2240 non-null   int64         
 2   education            2240 non-null   object        
 3   marital_status       2240 non-null   object        
 4   income               2240 non-null   float64       
 5   kidhome              2240 non-null   int64         
 6   teenhome             2240 non-null   int64         
 7   dt_customer          2240 non-null   datetime64[ns]
 8   recency              2240 non-null   int64         
 9   mntwines             2240 non-null   int64         
 10  mntfruits            2240 non-null   int64         
 11  mntmeatproducts      2240 non-null   int64         
 12  mntfishproducts      2240 non-null   int64         
 13  mntsweetproducts     2240 non-nul

In [146]:
df.head()

,id,year_birth,education,marital_status,income,kidhome,teenhome,dt_customer,recency,mntwines,...,numwebvisitsmonth,acceptedcmp3,acceptedcmp4,acceptedcmp5,acceptedcmp1,acceptedcmp2,complain,z_costcontact,z_revenue,response
0,5524,1957,Graduation,Single,58138.0,0,0,2012-09-04,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,2014-03-08,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,2013-08-21,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,2014-02-10,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,2014-01-19,94,173,...,5,0,0,0,0,0,0,3,11,0


In [147]:
df.shape

(2240, 29)

In [148]:
df.describe()

,id,year_birth,income,kidhome,teenhome,dt_customer,recency,mntwines,mntfruits,mntmeatproducts,...,numwebvisitsmonth,acceptedcmp3,acceptedcmp4,acceptedcmp5,acceptedcmp1,acceptedcmp2,complain,z_costcontact,z_revenue,response
count,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240,2240.000000,2240.000000,2240.000000,2240.000000,...,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.000000,2240.0,2240.0,2240.000000
mean,5592.159821,1968.805804,52237.975446,0.444196,0.506250,2013-07-10 10:01:42.857142784,49.109375,303.935714,26.302232,166.950000,...,5.316518,0.072768,0.074554,0.072768,0.064286,0.013393,0.009375,3.0,11.0,0.149107
min,0.000000,1893.000000,1730.000000,0.000000,0.000000,2012-07-30 00:00:00,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
25%,2828.250000,1959.000000,35538.750000,0.000000,0.000000,2013-01-16 00:00:00,24.000000,23.750000,1.000000,16.000000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
50%,5458.500000,1970.000000,51381.500000,0.000000,0.000000,2013-07-08 12:00:00,49.000000,173.500000,8.000000,67.000000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
75%,8427.750000,1977.000000,68289.750000,1.000000,1.000000,2013-12-30 06:00:00,74.000000,504.250000,33.000000,232.000000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.0,11.0,0.000000
max,11191.000000,1996.000000,666666.000000,2.000000,2.000000,2014-06-29 00:00:00,99.000000,1493.000000,199.000000,1725.000000,...,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,3.0,11.0,1.000000
std,3246.662198,11.984069,25037.955891,0.538398,0.544538,NaN,28.962453,336.597393,39.773434,225.715373,...,2.426645,0.259813,0.262728,0.259813,0.245316,0.114976,0.096391,0.0,0.0,0.356274


In [149]:
print("Final dataset shape:", df.shape)

print(
    "Total missing values:",
    df.isnull().sum().sum()
)

print(
    "Total duplicate rows:",
    df.duplicated().sum()
)

Final dataset shape: (2240, 29)
Total missing values: 0
Total duplicate rows: 0


In [150]:
# Final data quality summary

final_summary = pd.DataFrame({
    "Metric": [
        "Rows",
        "Columns",
        "Missing Income Values",
        "Total Missing Values",
        "Duplicate Rows",
        "Duplicate Customer IDs",
        "Invalid Dates"
    ],
    "Result": [
        df.shape[0],
        df.shape[1],
        df["income"].isnull().sum(),
        df.isnull().sum().sum(),
        df.duplicated().sum(),
        df["id"].duplicated().sum(),
        df["dt_customer"].isnull().sum()
    ]
})

final_summary

,Metric,Result
0,Rows,2240
1,Columns,29
2,Missing Income Values,0
3,Total Missing Values,0
4,Duplicate Rows,0
5,Duplicate Customer IDs,0
6,Invalid Dates,0


## Data Cleaning Summary

The raw Customer Personality Analysis dataset initially contained 2,240 rows and 29 columns.

The following preprocessing steps were performed:

1. Inspected the dataset structure, dimensions, and data types.
2. Identified missing values.
3. Found 24 missing values in the `Income` column.
4. Replaced missing `Income` values using the median income value of 51,381.5.
5. Checked for duplicate rows; no duplicate rows were found.
6. Checked for duplicate customer IDs; no duplicate IDs were found.
7. Standardized column names using lowercase letters and underscores.
8. Removed unnecessary whitespace from text fields.
9. Inspected categorical values for consistency.
10. Converted `Dt_Customer` to the `datetime64[ns]` data type.
11. Checked for invalid dates.
12. Validated important numerical columns.
13. Checked for negative values in relevant numerical fields.
14. Investigated potentially unusual birth-year values.
15. Performed final missing-value and duplicate checks.

The final dataset contains 2,240 rows and 29 columns and has no missing values or duplicate rows.

In [151]:
# Save the cleaned dataset

output_file = "/content/marketing_campaign_cleaned.csv"

df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully:")
print(output_file)

Cleaned dataset saved successfully:
/content/marketing_campaign_cleaned.csv


In [152]:
from google.colab import files

files.download("/content/marketing_campaign_cleaned.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>